synapse/rest/client/models.py

```diff
@@ -19,6 +20,12 @@
 from synapse.util.threepids import validate_email
 
 
+class ThreepidMedium(str, Enum):
+    # Per advice at https://pydantic-docs.helpmanual.io/usage/types/#enums-and-choices
```

Comment:
can you be specific about what advice you're applying / what this is commenting on?
The constants being lowercased perhaps (main thing that strikes me as weird)?

Reply:
Sure. My first thought was to say `medium: StrictStr` but I wanted to see if there was a way to say that "`medium` should be one of these specific strings". The link seems to be the blessed way to do this.

What I was suspicious of is the multiple inheritance from `str` and `Enum`. These `ThreepidMedium` types will get passed through to the rest of the application, where I want them to be treated like any other `str`. I'm not fully convinced that the rest of the application will be happy with that tbh... but I don't have anything more concrete than a bad feeling about this.

Having said that, looking [here](https://pydantic-docs.helpmanual.io/usage/types/#literal-type) it might be possible to write `medium: Literal["email", "msisdn"]`. If it works, that seems a bit less magic.

Reply:
> Having said that, looking [here](https://pydantic-docs.helpmanual.io/usage/types/#literal-type) it might be possible to write `medium: Literal["email", "msisdn"]`. If it works, that seems a bit less magic.

This seems to work pretty well:

```python
class ThreepidMediumEnumTestCase(stdlib_unittest.TestCase):
    class Model(BaseModel):
        medium: Literal["email", "msisdn"]

    def test_accepts_valid_medium_string(self) -> None:
        """Sanity check that Pydantic behaves sensibly with an enum-of-str

        This is arguably more of a test of a class that inherits from str and Enum
        simultaneously.
        """
        model = self.Model.parse_obj({"medium": "email"})
        self.assertIsInstance(model.medium, str)
        self.assertEqual(model.medium, "email")

    def test_rejects_invalid_medium_value(self) -> None:
        with self.assertRaises(ValidationError) as ctx:
            self.Model.parse_obj({"medium": "interpretive_dance"})
        print(ctx.exception)

    def test_rejects_invalid_medium_type(self) -> None:
        with self.assertRaises(ValidationError) as ctx:
            self.Model.parse_obj({"medium": 123})
        print(ctx.exception)
```

The error messages generated are

```
Ran 3 tests in 0.002s

PASSED (successes=3)

Process finished with exit code 0
1 validation error for Model
medium
  unexpected value; permitted: 'email', 'msisdn' (type=value_error.const; given=123; permitted=('email', 'msisdn'))
1 validation error for Model
medium
  unexpected value; permitted: 'email', 'msisdn' (type=value_error.const; given=interpretive_dance; permitted=('email', 'msisdn'))
```

(the errors are a little technical, but see #13337)

Reply:
I was actually totally fine with the enum, but enum constants usually have uppercase names — that's what I was mostly wondering about (alongside a slightly underspecified comment).

Reply:
Oh I see. That's fair.

I really want to like stdlib enums, but I've been bitten by their sharp corners in the past. We've even seen it ourselves with `IntEnum` a few times I think? In that light, I think I prefer the `Literal` approach. But I can also see that a proper `Enum` can be passed throughout the rest of the application.

Maybe we should dwell on this for a bit. (I'd be interested in other opinions too!)